---
title: 'A DADA2 workflow for Big Data: Paired-end (1.2)'
output: html_document
---

--------------------------

**<span style="color:red">More discussion of the computational strategy of the big data workflow is available in [the single-read version](bigdata_1_2.html).**

## Starting point

The workflow below expects demultiplexed, per-sample, gzipped fastq files for the primer-free forward reads of a Hiseq run to be in the directory `pathF` (defined below), and the corresponding primer-free reverse read files to be in a different directory `pathR`. The string parsing below expects forward and reverse filenames of the following format: `samplename_XXX.fastq.gz`, but other formats simply require modification to the filename parsing portions of these scripts.

**NOTE: The filename parsing in this workflow expects the paired forward and reverse fastq files to have the same `samplename_` prefix.**

## Filtering

Please consider the filtering parameters, one set of parameters is not optimal for all datasets.

**Filtering script:**
```{r filter, eval=FALSE}
library(dada2); packageVersion("dada2")
# File parsing
pathF <- "/path/to/FWD" # CHANGE ME to the directory containing your demultiplexed forward-read fastq files
pathR <- "/path/to/REV" # CHANGE ME, ...
filtpathF <- file.path(pathF, "filtered") # Filtered forward files go into the pathF/filtered/ subdirectory
filtpathR <- file.path(pathR, "filtered") # ...
fastqFs <- sort(list.files(pathF, pattern="fastq"))
fastqRs <- sort(list.files(pathR, pattern="fastq"))
dir.create(filtpathF)
dir.create(filtpathR)
if(length(fastqFs) != length(fastqRs)) stop("Forward and reverse files do not match.")
# Filtering: THESE PARAMETERS ARENT OPTIMAL FOR ALL DATASETS
for(i in seq_along(fastqFs)) {
  fqF <- fastqFs[[i]]; fqR <- fastqRs[[i]]
  fastqPairedFilter(c(file.path(pathF,fqF), file.path(pathR, fqR)),
              c(file.path(filtpathF, fqF), file.path(filtpathR, fqR)),
              truncLen=c(240,200),
              maxEE=2, truncQ=11, maxN=0, rm.phix=TRUE,
              compress=TRUE, verbose=TRUE)
}
```

## Sample inference

The `NSAM.LEARN` parameter below should be large enough that the total number of reads across `NSAM.LEARN` samples exceeds 1M. 

**Sample inference script:**
```{r inference, eval=FALSE}
library(dada2); packageVersion("dada2")
# File parsing
filtpathF <- "/path/to/FWD/filtered" # CHANGE ME to the directory containing your filtered forward fastq files
filtpathR <- "/path/to/FWD/filtered" # CHANGE ME ...
filtFs <- list.files(filtpathF, full.names = TRUE)
filtRs <- list.files(filtpathR, full.names = TRUE)
sample.names <- sapply(strsplit(basename(filtFs), "_"), `[`, 1) # Assumes filename = samplename_XXX.fastq.gz
sample.namesR <- sapply(strsplit(basename(filtRs), "_"), `[`, 1) # Assumes filename = samplename_XXX.fastq.gz
if(!identical(sample.names, sample.namesR)) stop("Forward and reverse files do not match.")
names(filtFs) <- sample.names
names(filtRs) <- sample.names
set.seed(100)
# Learn forward error rates
NSAM.LEARN <- 25 # Choose enough samples to have at least 1M reads
drp.learnF <- derepFastq(sample(filtFs, NSAM.LEARN))
dd.learnF <- dada(drp.learnF, err=NULL, selfConsist=TRUE, multithread=TRUE)
errF <- dd.learnF[[1]]$err_out
rm(drp.learnF);rm(dd.learnF)
# Learn reverse error rates
drp.learnR <- derepFastq(sample(filtRs, NSAM.LEARN))
dd.learnR <- dada(drp.learnR, err=NULL, selfConsist=TRUE, multithread=TRUE)
errR <- dd.learnR[[1]]$err_out
rm(drp.learnR);rm(dd.learnR)
# Sample inference and merger of paired-end reads
mergers <- vector("list", length(sample.names))
names(mergers) <- sample.names
for(sam in sample.names) {
  cat("Processing:", sam, "\n")
	derepF <- derepFastq(filtFs[[sam]])
	ddF <- dada(derepF, err=errF, multithread=TRUE)
	derepR <- derepFastq(filtRs[[sam]])
	ddR <- dada(derepR, err=errR, multithread=TRUE)
	merger <- mergePairs(ddF, derepF, ddR, derepR)
	mergers[[sam]] <- merger
}
rm(derepF); rm(derepR)
# Construct sequence table and remove chimeras
seqtab <- makeSequenceTable(mergers)
seqtab <- removeBimeraDenovo(seqtab, multithread=TRUE)
saveRDS(seqtab, "/path/to/run1/output/seqtab.rds") # CHANGE ME to where you want sequence table saved
```

The final result, the count matrix of samples (rows) by non-chimeric sequence variants (columns), is stored as as serialized R object. Read it back into R with `foo <- readRDS("/path/to/run1/output/seqtab.rds")`.
