<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />




<title>A DADA2 workflow for Big Data: Paired-end (1.4)</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/flatly.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-1.1/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-1.1/highlight.js"></script>
<link href="site_libs/font-awesome-4.5.0/css/font-awesome.min.css" rel="stylesheet" />
<link href="site_libs/ionicons-2.0.1/css/ionicons.min.css" rel="stylesheet" />

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs && document.readyState && document.readyState === "complete") {
   window.setTimeout(function() {
      hljs.initHighlighting();
   }, 0);
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>


</head>

<body>

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
button.code-folding-btn:focus {
  outline: none;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 60px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 65px;
  margin-top: -65px;
}

.section h2 {
  padding-top: 65px;
  margin-top: -65px;
}
.section h3 {
  padding-top: 65px;
  margin-top: -65px;
}
.section h4 {
  padding-top: 65px;
  margin-top: -65px;
}
.section h5 {
  padding-top: 65px;
  margin-top: -65px;
}
.section h6 {
  padding-top: 65px;
  margin-top: -65px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>


<div class="container-fluid main-container">

<!-- tabsets -->
<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});
</script>

<!-- code folding -->






<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">dada2</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="dada-installation.html">Install</a>
</li>
<li>
  <a href="tutorial.html">Tutorial</a>
</li>
<li>
  <a href="bigdata.html">Big Data</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    <span class="fa fa-gear"></span>
     
    Documentation
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="https://www.bioconductor.org/packages/3.5/bioc/manuals/dada2/man/dada2.pdf">Manual</a>
    </li>
    <li>
      <a href="assign.html">Taxonomy</a>
    </li>
    <li>
      <a href="pool.html">Pooling</a>
    </li>
    <li>
      <a href="faq.html">FAQ</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    <span class="fa fa-gear"></span>
     
    Evaluation
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="http://dx.doi.org/10.1038/nmeth.3869">Manuscript</a>
    </li>
    <li>
      <a href="SMBS_DADA2.pdf">Symposium Slides</a>
    </li>
    <li>
      <a href="SotA.html">Benchmarking</a>
    </li>
    <li>
      <a href="https://purl.stanford.edu/mh194vj6733">Manuscript code</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://www.twitter.com/bejcal">
    <span class="ion ion-social-twitter"></span>
     
  </a>
</li>
<li>
  <a href="https://github.com/benjjneb/dada2/issues">
    <span class="fa fa-question fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="https://github.com/benjjneb/dada2">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">A DADA2 workflow for Big Data: Paired-end (1.4)</h1>

</div>


<hr />
<p><strong><span style="color:red">More discussion of the computational strategy of the big data workflow is available in <a href="bigdata.html">the single-read version</a>.</strong></p>
<div id="starting-point" class="section level2">
<h2>Starting point</h2>
<p>The workflow below expects demultiplexed, per-sample, gzipped fastq files for the primer-free forward reads of a Hiseq run to be in the directory <code>pathF</code> (defined below), and the corresponding primer-free reverse read files to be in a different directory <code>pathR</code>. The string parsing below expects forward and reverse filenames of the following format: <code>samplename_XXX.fastq.gz</code>, but other formats simply require modification to the filename parsing portions of these scripts.</p>
<p><strong>NOTE: The filename parsing in this workflow expects the paired forward and reverse fastq files to have the same <code>samplename_</code> prefix.</strong></p>
</div>
<div id="filter" class="section level2">
<h2>Filter</h2>
<p>Please consider the filtering parameters, one set of parameters is not optimal for all datasets.</p>
<p><strong>Filtering script:</strong></p>
<pre class="r"><code>library(dada2); packageVersion(&quot;dada2&quot;)
# File parsing
pathF &lt;- &quot;/path/to/FWD&quot; # CHANGE ME to the directory containing your demultiplexed forward-read fastq files
pathR &lt;- &quot;/path/to/REV&quot; # CHANGE ME ...
filtpathF &lt;- file.path(pathF, &quot;filtered&quot;) # Filtered forward files go into the pathF/filtered/ subdirectory
filtpathR &lt;- file.path(pathR, &quot;filtered&quot;) # ...
fastqFs &lt;- sort(list.files(pathF, pattern=&quot;fastq.gz&quot;))
fastqRs &lt;- sort(list.files(pathR, pattern=&quot;fastq.gz&quot;))
if(length(fastqFs) != length(fastqRs)) stop(&quot;Forward and reverse files do not match.&quot;)
# Filtering: THESE PARAMETERS ARENT OPTIMAL FOR ALL DATASETS
filterAndTrim(fwd=file.path(pathF, fastqFs), filt=file.path(filtpathF, fastqFs),
              rev=file.path(pathR, fastqRs), filt.rev=file.path(filtpathR, fastqRs),
              truncLen=c(240,200), maxEE=2, truncQ=11, maxN=0, rm.phix=TRUE,
              compress=TRUE, verbose=TRUE, multithread=TRUE)</code></pre>
</div>
<div id="infer-sequence-variants" class="section level2">
<h2>Infer Sequence Variants</h2>
<p>This portion of the workflow should be performed on a run-by-run basis, as error rates can differ between runs.</p>
<p><strong>Sample inference script:</strong></p>
<pre class="r"><code>library(dada2); packageVersion(&quot;dada2&quot;)
# File parsing
filtpathF &lt;- &quot;/path/to/FWD/filtered&quot; # CHANGE ME to the directory containing your filtered forward fastq files
filtpathR &lt;- &quot;/path/to/FWD/filtered&quot; # CHANGE ME ...
filtFs &lt;- list.files(filtpathF, pattern=&quot;fastq.gz&quot;, full.names = TRUE)
filtRs &lt;- list.files(filtpathR, pattern=&quot;fastq.gz&quot;, full.names = TRUE)
sample.names &lt;- sapply(strsplit(basename(filtFs), &quot;_&quot;), `[`, 1) # Assumes filename = samplename_XXX.fastq.gz
sample.namesR &lt;- sapply(strsplit(basename(filtRs), &quot;_&quot;), `[`, 1) # Assumes filename = samplename_XXX.fastq.gz
if(!identical(sample.names, sample.namesR)) stop(&quot;Forward and reverse files do not match.&quot;)
names(filtFs) &lt;- sample.names
names(filtRs) &lt;- sample.names
set.seed(100)
# Learn forward error rates
errF &lt;- learnErrors(filtFs, nread=2e6, multithread=TRUE)
# Learn reverse error rates
errR &lt;- learnErrors(filtRs, nread=2e6, multithread=TRUE)
# Sample inference and merger of paired-end reads
mergers &lt;- vector(&quot;list&quot;, length(sample.names))
names(mergers) &lt;- sample.names
for(sam in sample.names) {
  cat(&quot;Processing:&quot;, sam, &quot;\n&quot;)
    derepF &lt;- derepFastq(filtFs[[sam]])
    ddF &lt;- dada(derepF, err=errF, multithread=TRUE)
    derepR &lt;- derepFastq(filtRs[[sam]])
    ddR &lt;- dada(derepR, err=errR, multithread=TRUE)
    merger &lt;- mergePairs(ddF, derepF, ddR, derepR)
    mergers[[sam]] &lt;- merger
}
rm(derepF); rm(derepR)
# Construct sequence table and remove chimeras
seqtab &lt;- makeSequenceTable(mergers)
saveRDS(seqtab, &quot;/path/to/run1/output/seqtab.rds&quot;) # CHANGE ME to where you want sequence table saved</code></pre>
<p>The final result, the count matrix of samples (rows) by non-chimeric sequence variants (columns), is stored as as serialized R object. Read it back into R with <code>foo &lt;- readRDS(&quot;/path/to/run1/output/seqtab.rds&quot;)</code>.</p>
</div>
<div id="merge-runs-remove-chimeras-assign-taxonomy" class="section level2">
<h2>Merge Runs, Remove Chimeras, Assign Taxonomy</h2>
<p>This part is the same as in the single-read version of this workflow.</p>
<p><strong>Chimera/Taxonomy script:</strong></p>
<pre class="r"><code>library(dada2); packageVersion(&quot;dada2&quot;)
# Merge multiple runs (if necessary)
st1 &lt;- readRDS(&quot;/path/to/run1/output/seqtab.rds&quot;)
st2 &lt;- readRDS(&quot;/path/to/run2/output/seqtab.rds&quot;)
st3 &lt;- readRDS(&quot;/path/to/run3/output/seqtab.rds&quot;)
st.all &lt;- mergeSequenceTables(st1, st2, st3)
# Remove chimeras
seqtab &lt;- removeBimeraDenovo(st.all, method=&quot;consensus&quot;, multithread=TRUE)
# Assign taxonomy
tax &lt;- assignTaxonomy(seqtab, &quot;/path/to/silva_nr_v123_train_set.fa.gz&quot;, multithread=TRUE)
# Write to disk
saveRDS(seqtab, &quot;/path/to/study/seqtab_final.rds&quot;) # CHANGE ME to where you want sequence table saved
saveRDS(tax, &quot;/pathto/study/tax_final.rds&quot;) # CHANGE ME ...</code></pre>
</div>

<div class="container" style="width: 100%;color:grey;text-align:right">
  <hr>
  Maintained by Benjamin Callahan (benjamin DOT j DOT callahan AT gmail DOT com)
</div>



</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
