<!DOCTYPE html>
<html>
  <head>
    <meta charset='utf-8'>
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <link href='https://fonts.googleapis.com/css?family=Chivo:900' rel='stylesheet' type='text/css'>
    <link rel="stylesheet" type="text/css" href="stylesheets/stylesheet.css" media="screen">
    <link rel="stylesheet" type="text/css" href="stylesheets/github-dark.css" media="screen">
    <link rel="stylesheet" type="text/css" href="stylesheets/print.css" media="print">
    <!--[if lt IE 9]>
    <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->
    <title>Dada2 by benjjneb</title>
  </head>

  <body>
    <div id="container">
      <div class="inner">

        <header>
          <h1>Dada2</h1>
          <h2>Fast, accurate, single-nucleotide resolution for amplicon data</h2>
        </header>

        <section id="downloads" class="clearfix">
          <a href="https://github.com/benjjneb/dada2/zipball/master" id="download-zip" class="button"><span>Download .zip</span></a>
          <a href="https://github.com/benjjneb/dada2/tarball/master" id="download-tar-gz" class="button"><span>Download .tar.gz</span></a>
          <a href="https://github.com/benjjneb/dada2" id="view-on-github" class="button"><span>View on GitHub</span></a>
        </section>

        <hr>

        <section id="main_content">

<h2>
<a id="dada2" class="anchor" href="#installation" aria-hidden="true"><span class="octicon octicon-link"></span></a>Installation</h2>

<p>See the <a href="R/dada-installation.html">DADA2 installation instructions.</a></p>

<h2>
<a id="dada2" class="anchor" href="#tutorial" aria-hidden="true"><span class="octicon octicon-link"></span></a>Tutorial</h2>

<p>See the <a href="R/tutorial.html">DADA2 walkthrough on paired end Illumina Miseq data.</a></p>

<h2>
<a id="dada2" class="anchor" href="#tutorial" aria-hidden="true"><span class="octicon octicon-link"></span></a>Advantages</h2>

<p>DADA2 infers exact sample sequences from Illumina amplicon data. Even if biological variants differ by 1 or 2 nucleotides, DADA2 will resolve them.</p>
  
<p>DADA2 reports fewer false sequences than other methods report false OTUs.</p>

<p>The computational costs of DADA2 scale linearly with the number of samples, and DADA2's memory requirements remain essentialy flat.
</p>

<h2>
<a id="dada2" class="anchor" href="#issues" aria-hidden="true"><span class="octicon octicon-link"></span></a>Issues</h2>

<p>Planned feature improvements are publicly catalogued at the main DADA2 development site on github; specifically on the "Issues" page for DADA2:</p>

<p><a href="https://github.com/benjjneb/dada2/issues">https://github.com/benjjneb/dada2/issues</a></p>

<p>If the feature you are hoping for is not listed, you are welcome to add it as a feature request "issue" on this page. This request will be publicly available and listed on the page.</p>

<p>Posts about bug reports and problems using DADA2 are also welcome on <a href="https://github.com/benjjneb/dada2/issues">the issue tracker.</a></p>

<h2><a id="dada2" class="anchor" href="#issues" aria-hidden="true"><span class="octicon octicon-link"></span></a>How?</h2>

<p></p>DADA2 makes simultaneous improvements over current methods in resolution, accuracy and performance scaling. How is this possible?</p>

<p>DADA2's crucial advantage is that it uses more of the data. The DADA2 error model incorporates quality information, which is ignored by all other methods after filtering. The DADA2 error model incorporates quantitative abundances, whereas most other methods use abundance ranks if they use abundance at all. The DADA2 error model identifies the differences between sequences, eg. A->C, whereas other methods merely count the mismatches. DADA2 uses an error model, whereas OTU methods compare Hamming distance to an arbitrary threshold. DADA2 can parameterize its error model from the data itself, rather than relying on previous datasets that may or may not reflect the PCR and sequencing protocols used in your study.</p>

<p>DADA2's performance gains come from the fact that it infers sequences exactly rather than constructing OTUs. Denovo OTUs cannot be compared across samples unless all samples were pooled during OTU construction. However, exact sequences **are** comparable across samples: If the sequences are the same, they are the same, if they are different, they are different. Thus DADA2 can analyze each sample independently, resulting in linear scaling with sample number and trivial parallelization.</p>

        </section>

        <footer>
          Dada2 is maintained by <a href="https://github.com/benjjneb">benjjneb</a><br>
          This page was generated by <a href="https://pages.github.com">GitHub Pages</a>. Tactile theme by <a href="https://twitter.com/jasonlong">Jason Long</a>.
        </footer>

        
      </div>
    </div>
  </body>
</html>
