---
title: "The DADA2 R package"
---

## Fast, accurate, single-nucleotide resolution for amplicon data

--------------------------

## Installation

Binaries for the current release version of DADA2 (1.0) are available from Bioconductor (note that you must have R 3.3 to install from Bioconductor!):

```S
## try http:// if https:// URLs are not supported
source("https://bioconductor.org/biocLite.R")
biocLite("dada2")
```

If you wish to install the latest and greatest version, or to install to R 3.2, see our [from-source installation instructions](dada-installation.html).

&nbsp;

## Tutorial

The [DADA2 walkthrough on paired end Illumina Miseq data](tutorial.html) goes through a typical example workflow.

You can also [assign taxonomy with dada2::assignTaxonomy(...)](assign.html).

&nbsp;

## Benchmarking

Our [manuscript](http://dx.doi.org/10.1038/nmeth.3869) compares the accuracy of DADA2 and other methods on several mock community datasets. Further [benchmarking of DADA2 against the methods evaluated in the recent QIIME benchmarking paper](SotA.html) is also available.

&nbsp;

## Advantages

* DADA2 infers exact sample sequences from Illumina amplicon data. Even if sequence variants differ by 1 or 2 nucleotides, DADA2 will resolve them.
  
* DADA2 reports fewer false sequences than other methods report false OTUs.

* The computational costs of DADA2 can scale linearly with the number of samples, and DADA2's memory requirements can remain essentialy flat.

* DADA2 is open source software, licensed under the [LGPL version 3](http://www.gnu.org/licenses/lgpl-3.0.en.html).

&nbsp;

## Support and Development

Planned feature improvements are publicly catalogued at the main DADA2 development site on github; specifically on the [issues tracker for DADA2](https://github.com/benjjneb/dada2/issues). If the feature you are hoping for is not listed, you are welcome to add it as a feature request on this page.

Bug reports and problems using DADA2 are also welcome on [the issues tracker](https://github.com/benjjneb/dada2/issues). We prefer posting to the issue tracker over email as these posts are searchable by other users who may experience the same problems.

&nbsp;

## How?

**Accurracy:** DADA2's crucial advantage is that it uses more of the data. The DADA2 error model incorporates quality information, which is ignored by all other methods after filtering. The DADA2 error model incorporates quantitative abundances, whereas most other methods use abundance ranks if they use abundance at all. The DADA2 error model identifies the differences between sequences, eg. A->C, whereas other methods merely count the mismatches. DADA2 can parameterize its error model from the data itself, rather than relying on previous datasets that may or may not reflect the PCR and sequencing protocols used in your study.

**Performance:** DADA2's performance gains come from the fact that it infers sequences exactly rather than constructing OTUs. Denovo OTUs cannot be compared across samples unless all samples were pooled during OTU construction. However, exact sequences **are** comparable across samples, as exact sequences are consistent labels. Thus DADA2 can analyze each sample independently, resulting in linear scaling with sample number and trivial parallelization.
