<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />




<title>The DADA2 R package</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/flatly.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-1.1/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-1.1/highlight.js"></script>
<link href="site_libs/font-awesome-4.5.0/css/font-awesome.min.css" rel="stylesheet" />
<link href="site_libs/ionicons-2.0.1/css/ionicons.min.css" rel="stylesheet" />

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs && document.readyState && document.readyState === "complete") {
   window.setTimeout(function() {
      hljs.initHighlighting();
   }, 0);
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>


</head>

<body>

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
button.code-folding-btn:focus {
  outline: none;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 60px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 65px;
  margin-top: -65px;
}

.section h2 {
  padding-top: 65px;
  margin-top: -65px;
}
.section h3 {
  padding-top: 65px;
  margin-top: -65px;
}
.section h4 {
  padding-top: 65px;
  margin-top: -65px;
}
.section h5 {
  padding-top: 65px;
  margin-top: -65px;
}
.section h6 {
  padding-top: 65px;
  margin-top: -65px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>


<div class="container-fluid main-container">

<!-- tabsets -->
<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});
</script>

<!-- code folding -->






<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">dada2</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="dada-installation.html">Install</a>
</li>
<li>
  <a href="tutorial.html">Tutorial</a>
</li>
<li>
  <a href="bigdata.html">Big Data</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    <span class="fa fa-gear"></span>
     
    Documentation
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="https://www.bioconductor.org/packages/3.5/bioc/manuals/dada2/man/dada2.pdf">Manual</a>
    </li>
    <li>
      <a href="assign.html">Taxonomy</a>
    </li>
    <li>
      <a href="pool.html">Pooling</a>
    </li>
    <li>
      <a href="faq.html">FAQ</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    <span class="fa fa-gear"></span>
     
    Evaluation
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="http://dx.doi.org/10.1038/nmeth.3869">Manuscript</a>
    </li>
    <li>
      <a href="SMBS_DADA2.pdf">Symposium Slides</a>
    </li>
    <li>
      <a href="SotA.html">Benchmarking</a>
    </li>
    <li>
      <a href="https://purl.stanford.edu/mh194vj6733">Manuscript code</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://www.twitter.com/bejcal">
    <span class="ion ion-social-twitter"></span>
     
  </a>
</li>
<li>
  <a href="https://github.com/benjjneb/dada2/issues">
    <span class="fa fa-question fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="https://github.com/benjjneb/dada2">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">The DADA2 R package</h1>

</div>


<div id="fast-accurate-single-nucleotide-resolution-for-amplicon-data" class="section level2">
<h2>Fast, accurate, single-nucleotide resolution for amplicon data</h2>
<hr />
</div>
<div id="installation" class="section level2">
<h2>Installation</h2>
<p>Binaries for the current release version of DADA2 (1.4) are available from Bioconductor (note that you must have R 3.4.0 or newer, and <a href="https://www.bioconductor.org/install/">Biocondcutor version 3.5</a>, to install the current release from Bioconductor):</p>
<pre class="s"><code>## try http:// if https:// URLs are not supported
source(&quot;https://bioconductor.org/biocLite.R&quot;)
biocLite(&quot;dada2&quot;)</code></pre>
<p>If you wish to install the latest and greatest development version, or to install to R 3.3, see our <a href="dada-installation.html">from-source installation instructions</a>.</p>
<p> </p>
</div>
<div id="tutorials" class="section level2">
<h2>Tutorials</h2>
<p><strong>Start here:</strong> The <a href="tutorial.html">DADA2 tutorial</a> goes through a typical workflow for paired end Illumina Miseq data: raw amplicon sequence data is processed into the table of biological sequence variants (SVs) present in each sample.</p>
<p>The <a href="bigdata.html">DADA2 Workflow on Big Data</a> goes through workflow optimized to run on large datasets (10s of millions to billions of reads).</p>
<p>Short demonstrations of <a href="assign.html">assigning taxonomy</a> and <a href="assign.html">assigning species</a> to sequences.</p>
<p> </p>
</div>
<div id="benchmarking" class="section level2">
<h2>Benchmarking</h2>
<p>Our <a href="http://dx.doi.org/10.1038/nmeth.3869">manuscript</a> (<a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4927377/">OA link</a>) compares the accuracy of DADA2 and other methods on several mock community datasets.</p>
<p>Further <a href="SotA.html">benchmarking of DADA2 against the methods evaluated in a recent QIIME1 benchmarking paper</a> is available.</p>
<p>We describe <a href="http://www.biorxiv.org/content/early/2017/03/07/113597">the broad advantages of exact sequence variants over OTUs in a recent preprint</a>.</p>
<p>And sometimes a picture says a thousand word:</p>
<div class="figure">
<img src="dada2_accuracy.jpg" />

</div>
<p> </p>
</div>
<div id="advantages" class="section level2">
<h2>Advantages</h2>
<ul>
<li><p><strong>High Resolution</strong>: DADA2 infers exact sequence variants from amplicon data, resolving differences of even 1 or 2 nucleotides.</p></li>
<li><p><strong>Accuracy</strong>: DADA2 reports fewer false positive sequence variants than other methods report false OTUs.</p></li>
<li><p><strong>Computational Scaling</strong>: The compute time of DADA2 scales linearly sample number, and memory requirements are essentially flat.</p></li>
<li><p><strong>Open Source</strong>: DADA2 is licensed under the <a href="http://www.gnu.org/licenses/lgpl-3.0.en.html">LGPL version 3</a>.</p></li>
</ul>
<p> </p>
</div>
<div id="support-and-development" class="section level2">
<h2>Support and Development</h2>
<p>Planned feature improvements are publicly catalogued at the main DADA2 development site on github; specifically on the <a href="https://github.com/benjjneb/dada2/issues">issues tracker for DADA2</a>. If the feature you are hoping for is not listed, you are welcome to add it as a feature request on this page.</p>
<p>Bug reports and problems using DADA2 are also welcome on <a href="https://github.com/benjjneb/dada2/issues">the issues tracker</a>. We prefer posting to the issue tracker over email as these posts are searchable by other users who may experience the same problems.</p>
<p> </p>
</div>
<div id="how" class="section level2">
<h2>How?</h2>
<p><strong>Accuracy:</strong> DADA2’s crucial advantage is that it uses more of the data. The DADA2 error model incorporates quality information, which is ignored by all other methods after filtering. The DADA2 error model incorporates quantitative abundances, whereas most other methods use abundance ranks if they use abundance at all. The DADA2 error model identifies the differences between sequences, eg. A-&gt;C, whereas other methods merely count the mismatches. DADA2 can parameterize its error model from the data itself, rather than relying on previous datasets that may or may not reflect the PCR and sequencing protocols used in your study.</p>
<p><strong>Performance:</strong> DADA2’s performance gains come from the fact that it infers sequences exactly rather than constructing OTUs. De novo OTUs cannot be compared across samples unless all samples were pooled during OTU construction. However, <a href="http://www.biorxiv.org/content/early/2017/03/07/113597">exact sequences <strong>are</strong> comparable across samples, as exact sequences are consistent labels</a>. Thus DADA2 can analyze each sample independently, resulting in linear scaling with sample number and trivial parallelization.</p>
</div>

<div class="container" style="width: 100%;color:grey;text-align:right">
  <hr>
  Maintained by Benjamin Callahan (benjamin DOT j DOT callahan AT gmail DOT com)
</div>



</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
