<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />




<title>DADA2 Pipeline Tutorial (1.4)</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/flatly.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-1.1/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-1.1/highlight.js"></script>
<link href="site_libs/font-awesome-4.5.0/css/font-awesome.min.css" rel="stylesheet" />
<link href="site_libs/ionicons-2.0.1/css/ionicons.min.css" rel="stylesheet" />

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs && document.readyState && document.readyState === "complete") {
   window.setTimeout(function() {
      hljs.initHighlighting();
   }, 0);
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>


</head>

<body>

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
button.code-folding-btn:focus {
  outline: none;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 60px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 65px;
  margin-top: -65px;
}

.section h2 {
  padding-top: 65px;
  margin-top: -65px;
}
.section h3 {
  padding-top: 65px;
  margin-top: -65px;
}
.section h4 {
  padding-top: 65px;
  margin-top: -65px;
}
.section h5 {
  padding-top: 65px;
  margin-top: -65px;
}
.section h6 {
  padding-top: 65px;
  margin-top: -65px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>


<div class="container-fluid main-container">

<!-- tabsets -->
<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});
</script>

<!-- code folding -->






<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">dada2</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="dada-installation.html">Install</a>
</li>
<li>
  <a href="tutorial.html">Tutorial</a>
</li>
<li>
  <a href="bigdata.html">Big Data</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    <span class="fa fa-gear"></span>
     
    Documentation
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="https://www.bioconductor.org/packages/3.5/bioc/manuals/dada2/man/dada2.pdf">Manual</a>
    </li>
    <li>
      <a href="assign.html">Taxonomy</a>
    </li>
    <li>
      <a href="pool.html">Pooling</a>
    </li>
    <li>
      <a href="faq.html">FAQ</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    <span class="fa fa-gear"></span>
     
    Evaluation
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="http://dx.doi.org/10.1038/nmeth.3869">Manuscript</a>
    </li>
    <li>
      <a href="SMBS_DADA2.pdf">Symposium Slides</a>
    </li>
    <li>
      <a href="SotA.html">Benchmarking</a>
    </li>
    <li>
      <a href="https://purl.stanford.edu/mh194vj6733">Manuscript code</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://www.twitter.com/bejcal">
    <span class="ion ion-social-twitter"></span>
     
  </a>
</li>
<li>
  <a href="https://github.com/benjjneb/dada2/issues">
    <span class="fa fa-question fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="https://github.com/benjjneb/dada2">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">DADA2 Pipeline Tutorial (1.4)</h1>

</div>


<p>Here we walk through version 1.4 of the DADA2 pipeline on a small multi-sample dataset. Our starting point is a set of Illumina-sequenced paired-end fastq files that have been split (or “demultiplexed”) by sample and from which the barcodes/adapters have already been removed. The end product is an <strong>amplicon sequence variant (ASV or SV) table</strong>, a higher-resolution analogue of the ubiquitous “OTU table”, which records the number of times each amplicon sequence variant was observed in each sample. We also assign taxonomy to the output sequences, and demonstrate how the data can be imported into the popular <a href="https://joey711.github.io/phyloseq/">phyloseq</a> R package for the analysis of microbiome data.</p>
<hr />
<div id="starting-point" class="section level1">
<h1>Starting point</h1>
<p>This workflow assumes that the data you are starting with meets certain criteria:</p>
<ul>
<li>Non-biological nucleotides have been removed (primers/adapters/barcodes…)</li>
<li>Samples are demultiplexed (split into individual per-sample fastqs)</li>
<li>If paired-end sequencing, the forward and reverse fastqs contain reads in matched order</li>
</ul>
<p>If these criteria are not true for your data (<strong>are you sure there aren’t any primers hanging around?</strong>) you need to remedy those issues before beginning this workflow. See <a href="faq.html">the FAQ</a> for some recommendations for common issues.</p>
</div>
<div id="getting-ready" class="section level1">
<h1>Getting ready</h1>
<p>First we load the <code>dada2</code> library. If you don’t already have the dada2 package, see the <a href="dada-installation.html">dada2 installation instructions</a>:</p>
<pre class="r"><code>library(dada2); packageVersion(&quot;dada2&quot;)</code></pre>
<pre><code>## [1] &#39;1.4.0&#39;</code></pre>
<p><em>If your dada2 version is &lt;1.3.4 you must upgrade to the current version to run this workflow. An <a href="tutorial_1_2.html">earlier version of this workflow for the 1.2 dada2 release</a> is also available.</em></p>
<p>The data we will work with are the same as those in the <a href="http://www.mothur.org/wiki/MiSeq_SOP">Mothur Miseq SOP</a> walkthrough. Download the <a href="http://www.mothur.org/w/images/d/d6/MiSeqSOPData.zip">example data</a> and unzip. These files represent longitudinal samples from a mouse post-weaning and one mock community control. For now just consider them paired-end fastq files to be processed. Define the following path variable so that it points to the extracted directory on <strong>your</strong> machine:</p>
<pre class="r"><code>path &lt;- &quot;~/MiSeq_SOP&quot; # CHANGE ME to the directory containing the fastq files after unzipping.
list.files(path)</code></pre>
<pre><code>##  [1] &quot;F3D0_S188_L001_R1_001.fastq&quot;   &quot;F3D0_S188_L001_R2_001.fastq&quot;  
##  [3] &quot;F3D1_S189_L001_R1_001.fastq&quot;   &quot;F3D1_S189_L001_R2_001.fastq&quot;  
##  [5] &quot;F3D141_S207_L001_R1_001.fastq&quot; &quot;F3D141_S207_L001_R2_001.fastq&quot;
##  [7] &quot;F3D142_S208_L001_R1_001.fastq&quot; &quot;F3D142_S208_L001_R2_001.fastq&quot;
##  [9] &quot;F3D143_S209_L001_R1_001.fastq&quot; &quot;F3D143_S209_L001_R2_001.fastq&quot;
## [11] &quot;F3D144_S210_L001_R1_001.fastq&quot; &quot;F3D144_S210_L001_R2_001.fastq&quot;
## [13] &quot;F3D145_S211_L001_R1_001.fastq&quot; &quot;F3D145_S211_L001_R2_001.fastq&quot;
## [15] &quot;F3D146_S212_L001_R1_001.fastq&quot; &quot;F3D146_S212_L001_R2_001.fastq&quot;
## [17] &quot;F3D147_S213_L001_R1_001.fastq&quot; &quot;F3D147_S213_L001_R2_001.fastq&quot;
## [19] &quot;F3D148_S214_L001_R1_001.fastq&quot; &quot;F3D148_S214_L001_R2_001.fastq&quot;
## [21] &quot;F3D149_S215_L001_R1_001.fastq&quot; &quot;F3D149_S215_L001_R2_001.fastq&quot;
## [23] &quot;F3D150_S216_L001_R1_001.fastq&quot; &quot;F3D150_S216_L001_R2_001.fastq&quot;
## [25] &quot;F3D2_S190_L001_R1_001.fastq&quot;   &quot;F3D2_S190_L001_R2_001.fastq&quot;  
## [27] &quot;F3D3_S191_L001_R1_001.fastq&quot;   &quot;F3D3_S191_L001_R2_001.fastq&quot;  
## [29] &quot;F3D5_S193_L001_R1_001.fastq&quot;   &quot;F3D5_S193_L001_R2_001.fastq&quot;  
## [31] &quot;F3D6_S194_L001_R1_001.fastq&quot;   &quot;F3D6_S194_L001_R2_001.fastq&quot;  
## [33] &quot;F3D7_S195_L001_R1_001.fastq&quot;   &quot;F3D7_S195_L001_R2_001.fastq&quot;  
## [35] &quot;F3D8_S196_L001_R1_001.fastq&quot;   &quot;F3D8_S196_L001_R2_001.fastq&quot;  
## [37] &quot;F3D9_S197_L001_R1_001.fastq&quot;   &quot;F3D9_S197_L001_R2_001.fastq&quot;  
## [39] &quot;filtered&quot;                      &quot;HMP_MOCK.v35.fasta&quot;           
## [41] &quot;Mock_S280_L001_R1_001.fastq&quot;   &quot;Mock_S280_L001_R2_001.fastq&quot;  
## [43] &quot;mouse.dpw.metadata&quot;            &quot;mouse.time.design&quot;            
## [45] &quot;stability.batch&quot;               &quot;stability.files&quot;              
## [47] &quot;test&quot;</code></pre>
<p>If the package successfully loaded and your listed files match those here, you are ready to go through the DADA2 pipeline.</p>
<p> </p>
</div>
<div id="filter-and-trim" class="section level1">
<h1>Filter and Trim</h1>
<p>First we read in the names of the fastq files, and perform some string manipulation to get lists of the forward and reverse fastq files in matched order:</p>
<pre class="r"><code># Sort ensures forward/reverse reads are in same order
fnFs &lt;- sort(list.files(path, pattern=&quot;_R1_001.fastq&quot;))
fnRs &lt;- sort(list.files(path, pattern=&quot;_R2_001.fastq&quot;))
# Extract sample names, assuming filenames have format: SAMPLENAME_XXX.fastq
sample.names &lt;- sapply(strsplit(fnFs, &quot;_&quot;), `[`, 1)
# Specify the full path to the fnFs and fnRs
fnFs &lt;- file.path(path, fnFs)
fnRs &lt;- file.path(path, fnRs)</code></pre>
<div style="border: 1px solid red;padding: 5px;background-color: #fff6f6;">
<strong><span style="color:red">If using this workflow on your own data:</span></strong> The string manipulations may have to be modified, especially the extraction of sample names from the file names.
</div>
<p> </p>
<div id="examine-quality-profiles-of-forward-and-reverse-reads" class="section level2">
<h2>Examine quality profiles of forward and reverse reads</h2>
<p>It is important to look at your data. We start by visualizing the quality profiles of the forward reads:</p>
<pre class="r"><code>plotQualityProfile(fnFs[1:2])</code></pre>
<p><img src="tutorial_files/figure-html/see-quality-F-1.png" width="672" /></p>
<p>The forward reads are good quality. We generally advise trimming the last few nucleotides to avoid less well-controlled errors that can arise there. There is no suggestion from these quality profiles that any additional trimming is needed, so we will truncate the forward reads at position 240 (trimming the last 10 nucleotides).</p>
<p>Now we visualize the quality profile of the reverse reads:</p>
<pre class="r"><code>plotQualityProfile(fnRs[1:2])</code></pre>
<p><img src="tutorial_files/figure-html/see-quality-R-1.png" width="672" /></p>
<p>The reverse reads are significantly worse quality, especially at the end, which is common in Illumina sequencing. This isn’t too worrisome, DADA2 incorporates quality information into its error model which makes the algorithm <a href="https://twitter.com/bejcal/status/771010634074820608">robust to lower quality sequence</a>, but trimming as the average qualities crash is still a good idea as long as our reads will still overlap. We will truncate at position 160 where the quality distribution crashes.</p>
<div style="border: 1px solid red;padding: 5px;background-color: #fff6f6;">
<strong><span style="color:red">If using this workflow on your own data:</span></strong> <strong>Your reads must overlap after truncation in order to merge them later!!!</strong> The tutorial is using 2x250 V4 sequence data, so the forward and reverse reads almost completely overlap and our trimming can be completely guided by the quality scores. If you are using a less-overlapping primer set, like V1-V2 or V3-V4, your <code>truncLen</code> must be large enough to maintain the overlap between them (the more the better).
</div>
<p> </p>
</div>
<div id="perform-filtering-and-trimming" class="section level2">
<h2>Perform filtering and trimming</h2>
<p>We define the filenames for the filtered fastq.gz files:</p>
<pre class="r"><code>filt_path &lt;- file.path(path, &quot;filtered&quot;) # Place filtered files in filtered/ subdirectory
filtFs &lt;- file.path(filt_path, paste0(sample.names, &quot;_F_filt.fastq.gz&quot;))
filtRs &lt;- file.path(filt_path, paste0(sample.names, &quot;_R_filt.fastq.gz&quot;))</code></pre>
<p>We’ll use standard filtering parameters: <code>maxN=0</code> (DADA2 requires no Ns), <code>truncQ=2</code>, <code>rm.phix=TRUE</code> and <code>maxEE=2</code>. The <code>maxEE</code> parameter sets the maximum number of “expected errors” allowed in a read, which is <a href="http://www.drive5.com/usearch/manual/expected_errors.html">a better filter than simply averaging quality scores</a>.</p>
<p><strong>Filter the forward and reverse reads</strong>:</p>
<pre class="r"><code>out &lt;- filterAndTrim(fnFs, filtFs, fnRs, filtRs, truncLen=c(240,160),
              maxN=0, maxEE=c(2,2), truncQ=2, rm.phix=TRUE,
              compress=TRUE, multithread=TRUE)
head(out)</code></pre>
<pre><code>##                               reads.in reads.out
## F3D0_S188_L001_R1_001.fastq       7793      7113
## F3D1_S189_L001_R1_001.fastq       5869      5299
## F3D141_S207_L001_R1_001.fastq     5958      5463
## F3D142_S208_L001_R1_001.fastq     3183      2914
## F3D143_S209_L001_R1_001.fastq     3178      2941
## F3D144_S210_L001_R1_001.fastq     4827      4312</code></pre>
<div style="border: 1px solid red;padding: 5px;background-color: #fff6f6;">
<strong><span style="color:red">If using this workflow on your own data:</span></strong> The standard filtering parameters are starting points, not set in stone. For example, if too few reads are passing the filter, considering relaxing <code>maxEE</code>, perhaps especially on the reverse reads (eg. <code>maxEE=c(2,5)</code>). If you want to speed up downstream computation, consider tightening <code>maxEE</code>. For paired-end reads consider the length of your amplicon when choosing <code>truncLen</code> as your reads must overlap after truncation in order to merge them later.
</div>
<div style="border: 1px solid red;padding: 5px;background-color: #fff6f6;margin-top: 15px;">
<strong><span style="color:red">If using this workflow on your own data:</span></strong> For common ITS amplicon strategies, it is undesirable to truncate reads to a fixed length due to the large amount of length variation at that locus. That is OK, just leave out <code>truncLen</code>. Make sure you removed the forward and reverse primers from both the forward and reverse reads though!
</div>
<p> </p>
</div>
</div>
<div id="learn-the-error-rates" class="section level1">
<h1>Learn the Error Rates</h1>
<p>The DADA2 algorithm depends on a parametric error model (<code>err</code>) and every amplicon dataset has a different set of error rates. The <code>learnErrors</code> method learns the error model from the data, by alternating estimation of the error rates and inference of sample composition until they converge on a jointly consistent solution. As in many optimization problems, the algorithm must begin with an initial guess, for which the maximum possible error rates in this data are used (the error rates if only the most abundant sequence is correct and all the rest are errors).</p>
<p><em>The following runs in about 4 minutes on a 2013 Macbook Pro:</em></p>
<pre class="r"><code>errF &lt;- learnErrors(filtFs, multithread=TRUE)</code></pre>
<pre><code>## Initializing error rates to maximum possible estimate.
## Sample 1 - 7113 reads in 1979 unique sequences.
## Sample 2 - 5299 reads in 1639 unique sequences.
## Sample 3 - 5463 reads in 1477 unique sequences.
## Sample 4 - 2914 reads in 904 unique sequences.
## Sample 5 - 2941 reads in 939 unique sequences.
## Sample 6 - 4312 reads in 1267 unique sequences.
## Sample 7 - 6741 reads in 1756 unique sequences.
## Sample 8 - 4560 reads in 1438 unique sequences.
## Sample 9 - 15637 reads in 3590 unique sequences.
## Sample 10 - 11413 reads in 2762 unique sequences.
## Sample 11 - 12017 reads in 3021 unique sequences.
## Sample 12 - 5032 reads in 1566 unique sequences.
## Sample 13 - 18075 reads in 3707 unique sequences.
## Sample 14 - 6250 reads in 1479 unique sequences.
## Sample 15 - 4052 reads in 1195 unique sequences.
## Sample 16 - 7369 reads in 1832 unique sequences.
## Sample 17 - 4765 reads in 1183 unique sequences.
## Sample 18 - 4871 reads in 1382 unique sequences.
## Sample 19 - 6504 reads in 1709 unique sequences.
## Sample 20 - 4314 reads in 897 unique sequences.
##    selfConsist step 2 
##    selfConsist step 3 
##    selfConsist step 4 
##    selfConsist step 5 
## 
## 
## Convergence after  5  rounds.
## Total reads used:  139642</code></pre>
<pre class="r"><code>errR &lt;- learnErrors(filtRs, multithread=TRUE)</code></pre>
<pre><code>## Initializing error rates to maximum possible estimate.
## Sample 1 - 7113 reads in 1660 unique sequences.
## Sample 2 - 5299 reads in 1349 unique sequences.
## Sample 3 - 5463 reads in 1335 unique sequences.
## Sample 4 - 2914 reads in 853 unique sequences.
## Sample 5 - 2941 reads in 880 unique sequences.
## Sample 6 - 4312 reads in 1286 unique sequences.
## Sample 7 - 6741 reads in 1803 unique sequences.
## Sample 8 - 4560 reads in 1265 unique sequences.
## Sample 9 - 15637 reads in 3414 unique sequences.
## Sample 10 - 11413 reads in 2522 unique sequences.
## Sample 11 - 12017 reads in 2771 unique sequences.
## Sample 12 - 5032 reads in 1415 unique sequences.
## Sample 13 - 18075 reads in 3290 unique sequences.
## Sample 14 - 6250 reads in 1390 unique sequences.
## Sample 15 - 4052 reads in 1134 unique sequences.
## Sample 16 - 7369 reads in 1635 unique sequences.
## Sample 17 - 4765 reads in 1084 unique sequences.
## Sample 18 - 4871 reads in 1161 unique sequences.
## Sample 19 - 6504 reads in 1502 unique sequences.
## Sample 20 - 4314 reads in 732 unique sequences.
##    selfConsist step 2 
##    selfConsist step 3 
##    selfConsist step 4 
##    selfConsist step 5 
##    selfConsist step 6 
## 
## 
## Convergence after  6  rounds.
## Total reads used:  139642</code></pre>
<p>It is always worthwhile, as a sanity check if nothing else, to visualize the estimated error rates:</p>
<pre class="r"><code>plotErrors(errF, nominalQ=TRUE)</code></pre>
<p><img src="tutorial_files/figure-html/plot-errors-1.png" width="672" /></p>
<p>The error rates for each possible transition (eg. A-&gt;C, A-&gt;G, …) are shown. Points are the observed error rates for each consensus quality score. The black line shows the estimated error rates after convergence. The red line shows the error rates expected under the nominal definition of the Q-value. Here the black line (the estimated rates) fits the observed rates well, and the error rates drop with increased quality as expected. Everything looks reasonable and we proceed with confidence.</p>
<div style="border: 1px solid red;padding: 5px;background-color: #fff6f6;">
<strong><span style="color:red">If using this workflow on your own data:</span></strong> Parameter learning is computationally intensive, so by default the <code>learnErrors</code> function uses only a subset of the data (the first 1M reads). If the plotted error model does not look like a good fit, try increasing the <code>nreads</code> parameter to see if the fit improves.
</div>
<p> </p>
</div>
<div id="dereplication" class="section level1">
<h1>Dereplication</h1>
<p>Dereplication combines all identical sequencing reads into into “unique sequences” with a corresponding “abundance”: the number of reads with that unique sequence. Dereplication substantially reduces computation time by eliminating redundant comparisons.</p>
<p>Dereplication in the DADA2 pipeline has one crucial addition from other pipelines: <strong>DADA2 retains a summary of the quality information associated with each unique sequence</strong>. The consensus quality profile of a unique sequence is the average of the positional qualities from the dereplicated reads. These quality profiles inform the error model of the subsequent denoising step, significantly increasing DADA2’s accuracy.</p>
<p><strong>Dereplicate the filtered fastq files</strong>:</p>
<pre class="r"><code>derepFs &lt;- derepFastq(filtFs, verbose=TRUE)
derepRs &lt;- derepFastq(filtRs, verbose=TRUE)
# Name the derep-class objects by the sample names
names(derepFs) &lt;- sample.names
names(derepRs) &lt;- sample.names</code></pre>
<div style="border: 1px solid red;padding: 5px;background-color: #fff6f6;">
<strong><span style="color:red">If using this workflow on your own data:</span></strong> The tutorial dataset is small enough to easily load into memory. If your dataset exceeds available RAM, it is preferable to process samples one-by-one in a streaming fashion: see the <a href="bigdata.html">DADA2 Workflow on Big Data</a> for an example.
</div>
<p> </p>
</div>
<div id="sample-inference" class="section level1">
<h1>Sample Inference</h1>
<p>We are now ready to apply the core sequence-variant inference algorithm to the dereplicated data.</p>
<p><strong>Infer the sequence variants in each sample</strong>:</p>
<pre class="r"><code>dadaFs &lt;- dada(derepFs, err=errF, multithread=TRUE)</code></pre>
<pre><code>## Sample 1 - 7113 reads in 1979 unique sequences.
## Sample 2 - 5299 reads in 1639 unique sequences.
## Sample 3 - 5463 reads in 1477 unique sequences.
## Sample 4 - 2914 reads in 904 unique sequences.
## Sample 5 - 2941 reads in 939 unique sequences.
## Sample 6 - 4312 reads in 1267 unique sequences.
## Sample 7 - 6741 reads in 1756 unique sequences.
## Sample 8 - 4560 reads in 1438 unique sequences.
## Sample 9 - 15637 reads in 3590 unique sequences.
## Sample 10 - 11413 reads in 2762 unique sequences.
## Sample 11 - 12017 reads in 3021 unique sequences.
## Sample 12 - 5032 reads in 1566 unique sequences.
## Sample 13 - 18075 reads in 3707 unique sequences.
## Sample 14 - 6250 reads in 1479 unique sequences.
## Sample 15 - 4052 reads in 1195 unique sequences.
## Sample 16 - 7369 reads in 1832 unique sequences.
## Sample 17 - 4765 reads in 1183 unique sequences.
## Sample 18 - 4871 reads in 1382 unique sequences.
## Sample 19 - 6504 reads in 1709 unique sequences.
## Sample 20 - 4314 reads in 897 unique sequences.</code></pre>
<pre class="r"><code>dadaRs &lt;- dada(derepRs, err=errR, multithread=TRUE)</code></pre>
<pre><code>## Sample 1 - 7113 reads in 1660 unique sequences.
## Sample 2 - 5299 reads in 1349 unique sequences.
## Sample 3 - 5463 reads in 1335 unique sequences.
## Sample 4 - 2914 reads in 853 unique sequences.
## Sample 5 - 2941 reads in 880 unique sequences.
## Sample 6 - 4312 reads in 1286 unique sequences.
## Sample 7 - 6741 reads in 1803 unique sequences.
## Sample 8 - 4560 reads in 1265 unique sequences.
## Sample 9 - 15637 reads in 3414 unique sequences.
## Sample 10 - 11413 reads in 2522 unique sequences.
## Sample 11 - 12017 reads in 2771 unique sequences.
## Sample 12 - 5032 reads in 1415 unique sequences.
## Sample 13 - 18075 reads in 3290 unique sequences.
## Sample 14 - 6250 reads in 1390 unique sequences.
## Sample 15 - 4052 reads in 1134 unique sequences.
## Sample 16 - 7369 reads in 1635 unique sequences.
## Sample 17 - 4765 reads in 1084 unique sequences.
## Sample 18 - 4871 reads in 1161 unique sequences.
## Sample 19 - 6504 reads in 1502 unique sequences.
## Sample 20 - 4314 reads in 732 unique sequences.</code></pre>
<p>Inspecting the dada-class object returned by dada:</p>
<pre class="r"><code>dadaFs[[1]]</code></pre>
<pre><code>## dada-class: object describing DADA2 denoising results
## 128 sample sequences were inferred from 1979 input unique sequences.
## Key parameters: OMEGA_A = 1e-40, BAND_SIZE = 16, USE_QUALS = TRUE</code></pre>
<p>The DADA2 algorithm inferred 128 real sequence variants from the 1979 unique sequences in the first sample. There is much more to the <code>dada-class</code> return object than this (see <code>help(&quot;dada-class&quot;)</code> for some info), including multiple diagnostics about the quality of each inferred sequence variant, but that is beyond the scope of an introductory tutorial.</p>
<div style="border: 1px solid red;padding: 5px;background-color: #fff6f6;">
<strong><span style="color:red">If using this workflow on your own data:</span></strong> All samples are simultaneously loaded into memory in the tutorial. If you are dealing with datasets that approach or exceed available RAM, it is preferable to process samples one-by-one in a streaming fashion: see the <strong><a href="bigdata.html">DADA2 Workflow on Big Data</a></strong> for an example.
</div>
<div style="border: 1px solid red;padding: 5px;background-color: #fff6f6;margin-top: 15px;">
<strong><span style="color:red">If using this workflow on your own data:</span></strong> By default, the <code>dada</code> function processes each sample independently, but pooled processing is available with <code>pool=TRUE</code> and that may give better results for low sampling depths at the cost of increased computation time. See our <a href="pool.html">discussion about pooling samples for sample inference</a>.
</div>
<div style="border: 1px solid red;padding: 5px;background-color: #fff6f6;margin-top: 15px;">
<strong><span style="color:red">If using this workflow on your own data:</span></strong> DADA2 also supports 454 and Ion Torrent data, but <a href="faq.html#can-i-use-dada2-with-my-454-or-ion-torrent-data">we recommend some minor parameter changes</a> for those sequencing technologies. The adventurous can explore <code>?setDadaOpt</code> for other adjustable algorithm parameters.
</div>
<p> </p>
</div>
<div id="merge-paired-reads" class="section level1">
<h1>Merge paired reads</h1>
<p>Spurious sequence variants are further reduced by merging overlapping reads. The core function here is <code>mergePairs</code>, which depends on the forward and reverse reads being in matching order at the time they were dereplicated.</p>
<p><strong>Merge the denoised forward and reverse reads</strong>:</p>
<pre class="r"><code>mergers &lt;- mergePairs(dadaFs, derepFs, dadaRs, derepRs, verbose=TRUE)
# Inspect the merger data.frame from the first sample
head(mergers[[1]])</code></pre>
<pre><code>##                                                                                                                                                                                                                                                       sequence
## 1 TACGGAGGATGCGAGCGTTATCCGGATTTATTGGGTTTAAAGGGTGCGCAGGCGGAAGATCAAGTCAGCGGTAAAATTGAGAGGCTCAACCTCTTCGAGCCGTTGAAACTGGTTTTCTTGAGTGAGCGAGAAGTATGCGGAATGCGTGGTGTAGCGGTGAAATGCATAGATATCACGCAGAACTCCGATTGCGAAGGCAGCATACCGGCGCTCAACTGACGCTCATGCACGAAAGTGTGGGTATCGAACAGG
## 2 TACGGAGGATGCGAGCGTTATCCGGATTTATTGGGTTTAAAGGGTGCGTAGGCGGCCTGCCAAGTCAGCGGTAAAATTGCGGGGCTCAACCCCGTACAGCCGTTGAAACTGCCGGGCTCGAGTGGGCGAGAAGTATGCGGAATGCGTGGTGTAGCGGTGAAATGCATAGATATCACGCAGAACCCCGATTGCGAAGGCAGCATACCGGCGCCCTACTGACGCTGAGGCACGAAAGTGCGGGGATCAAACAGG
## 3 TACGGAGGATGCGAGCGTTATCCGGATTTATTGGGTTTAAAGGGTGCGTAGGCGGGCTGTTAAGTCAGCGGTCAAATGTCGGGGCTCAACCCCGGCCTGCCGTTGAAACTGGCGGCCTCGAGTGGGCGAGAAGTATGCGGAATGCGTGGTGTAGCGGTGAAATGCATAGATATCACGCAGAACTCCGATTGCGAAGGCAGCATACCGGCGCCCGACTGACGCTGAGGCACGAAAGCGTGGGTATCGAACAGG
## 4 TACGGAGGATGCGAGCGTTATCCGGATTTATTGGGTTTAAAGGGTGCGTAGGCGGGCTTTTAAGTCAGCGGTAAAAATTCGGGGCTCAACCCCGTCCGGCCGTTGAAACTGGGGGCCTTGAGTGGGCGAGAAGAAGGCGGAATGCGTGGTGTAGCGGTGAAATGCATAGATATCACGCAGAACCCCGATTGCGAAGGCAGCCTTCCGGCGCCCTACTGACGCTGAGGCACGAAAGTGCGGGGATCGAACAGG
## 5 TACGGAGGATGCGAGCGTTATCCGGATTTATTGGGTTTAAAGGGTGCGCAGGCGGACTCTCAAGTCAGCGGTCAAATCGCGGGGCTCAACCCCGTTCCGCCGTTGAAACTGGGAGCCTTGAGTGCGCGAGAAGTAGGCGGAATGCGTGGTGTAGCGGTGAAATGCATAGATATCACGCAGAACTCCGATTGCGAAGGCAGCCTACCGGCGCGCAACTGACGCTCATGCACGAAAGCGTGGGTATCGAACAGG
## 6 TACGGAGGATGCGAGCGTTATCCGGATTTATTGGGTTTAAAGGGTGCGTAGGCGGGATGCCAAGTCAGCGGTAAAAAAGCGGTGCTCAACGCCGTCGAGCCGTTGAAACTGGCGTTCTTGAGTGGGCGAGAAGTATGCGGAATGCGTGGTGTAGCGGTGAAATGCATAGATATCACGCAGAACTCCGATTGCGAAGGCAGCATACCGGCGCCCTACTGACGCTGAGGCACGAAAGCGTGGGTATCGAACAGG
##   abundance forward reverse nmatch nmismatch nindel prefer accept
## 1       586       1       1    148         0      0      1   TRUE
## 2       471       2       2    148         0      0      2   TRUE
## 3       451       3       4    148         0      0      1   TRUE
## 4       433       4       3    148         0      0      2   TRUE
## 5       353       5       6    148         0      0      1   TRUE
## 6       285       6       5    148         0      0      2   TRUE</code></pre>
<p>We now have a <code>data.frame</code> for each sample with the merged <code>$sequence</code>, its <code>$abundance</code>, and the indices of the merged <code>$forward</code> and <code>$reverse</code> denoised sequences. Paired reads that did not exactly overlap were removed by <code>mergePairs</code>.</p>
<div style="border: 1px solid red;padding: 5px;background-color: #fff6f6;">
<strong><span style="color:red">If using this workflow on your own data:</span></strong> Most of your <strong>reads</strong> should successfully merge. If that is not the case upstream parameters may need to be revisited: Did you trim away the overlap between your reads?
</div>
<p> </p>
</div>
<div id="construct-sequence-table" class="section level1">
<h1>Construct sequence table</h1>
<p>We can now construct a “sequence table” of our mouse samples, a higher-resolution version of the “OTU table” produced by classical methods:</p>
<pre class="r"><code>seqtab &lt;- makeSequenceTable(mergers)</code></pre>
<pre><code>## The sequences being tabled vary in length.</code></pre>
<pre class="r"><code>dim(seqtab)</code></pre>
<pre><code>## [1]  20 288</code></pre>
<pre class="r"><code># Inspect distribution of sequence lengths
table(nchar(getSequences(seqtab)))</code></pre>
<pre><code>## 
## 251 252 253 254 255 
##   1  87 192   6   2</code></pre>
<p>The sequence table is a <code>matrix</code> with rows corresponding to (and named by) the samples, and columns corresponding to (and named by) the sequence variants. The lengths of our merged sequences all fall within the expected range for this V4 amplicon.</p>
<div style="border: 1px solid red;padding: 5px;background-color: #fff6f6;">
<strong><span style="color:red">If using this workflow on your own data:</span></strong> Sequences that are much longer or shorter than expected may be the result of non-specific priming, and may be worth removing (eg. <code>seqtab2 &lt;- seqtab[,nchar(colnames(seqtab)) %in% seq(250,256)]</code>). This is analogous to “cutting a band” in-silico to get amplicons of the targeted length.
</div>
<p> </p>
</div>
<div id="remove-chimeras" class="section level1">
<h1>Remove chimeras</h1>
<p>The core <code>dada</code> method removes substitution and indel errors, but chimeras remain. Fortunately, the accuracy of the sequences after denoising makes identifying chimeras simpler than it is when dealing with fuzzy OTUs: all sequences which can be exactly reconstructed as a bimera (two-parent chimera) from more abundant sequences.</p>
<p><strong>Remove chimeric sequences</strong>:</p>
<pre class="r"><code>seqtab.nochim &lt;- removeBimeraDenovo(seqtab, method=&quot;consensus&quot;, multithread=TRUE, verbose=TRUE)
dim(seqtab.nochim)</code></pre>
<pre><code>## [1]  20 229</code></pre>
<pre class="r"><code>sum(seqtab.nochim)/sum(seqtab)</code></pre>
<pre><code>## [1] 0.9643192</code></pre>
<p>The fraction of chimeras varies based on factors including experimental procedures and sample complexity, but can be substantial. Here chimeras make up about 20% of the inferred sequence variants, but those variants account for only about 4% of the total sequence reads.</p>
<div style="border: 1px solid red;padding: 5px;background-color: #fff6f6;">
<strong><span style="color:red">If using this workflow on your own data:</span></strong> Most of your <strong>reads</strong> should remain after chimera removal (it is not uncommon for a majority of <strong>sequence variants</strong> to be removed though). If most of your reads were removed as chimeric, upstream processing may need to be revisited. In almost all cases this is caused by primer sequences with ambiguous nucleotides that were not removed prior to beginning the DADA2 pipeline.
</div>
<p> </p>
</div>
<div id="track-reads-through-the-pipeline" class="section level1">
<h1>Track reads through the pipeline</h1>
<p>As a final check of our progress, we’ll look at the number of reads that made it through each step in the pipeline:</p>
<pre class="r"><code>getN &lt;- function(x) sum(getUniques(x))
track &lt;- cbind(out, sapply(dadaFs, getN), sapply(mergers, getN), rowSums(seqtab), rowSums(seqtab.nochim))
colnames(track) &lt;- c(&quot;input&quot;, &quot;filtered&quot;, &quot;denoised&quot;, &quot;merged&quot;, &quot;tabled&quot;, &quot;nonchim&quot;)
rownames(track) &lt;- sample.names
head(track)</code></pre>
<pre><code>##        input filtered denoised merged tabled nonchim
## F3D0    7793     7113     7113   6600   6600    6588
## F3D1    5869     5299     5299   5078   5078    5067
## F3D141  5958     5463     5463   5047   5047    4928
## F3D142  3183     2914     2914   2663   2663    2600
## F3D143  3178     2941     2941   2575   2575    2550
## F3D144  4827     4312     4312   3668   3668    3527</code></pre>
<p>Looks good, we kept the majority of our raw reads, and there is no over-large drop associated with any single step.</p>
<div style="border: 1px solid red;padding: 5px;background-color: #fff6f6;">
<strong><span style="color:red">If using this workflow on your own data:</span></strong> This is a great place to do a last <strong>sanity check</strong>. Outside of filtering (depending on how stringent you want to be) there should no step in which a majority of reads are lost. If a majority of reads failed to merge, you may need to revisit the <code>truncLen</code> parameter used in the filtering step and make sure that the truncated reads span your amplicon. If a majority of reads failed to pass the chimera check, you may need to revisit the removal of primers, as the ambiguous nucleotides in unremoved primers interfere with chimera identification.
</div>
<p> </p>
</div>
<div id="assign-taxonomy" class="section level1">
<h1>Assign taxonomy</h1>
<p>It is common at this point, especially in 16S/18S/ITS amplicon sequencing, to classify sequence variants taxonomically. The DADA2 package provides a native implementation of <a href="http://www.ncbi.nlm.nih.gov/pubmed/17586664">the RDP’s naive Bayesian classifier</a> for this purpose. The <code>assignTaxonomy</code> function takes a set of sequences and a training set of taxonomically classified sequences, and outputs the taxonomic assignments with at least <code>minBoot</code> bootstrap confidence.</p>
<p>We maintain <a href="training.html">formatted training fastas for the RDP training set, GreenGenes clustered at 97% identity, and the Silva reference database</a>. For fungal taxonomy, the General Fasta release files from the <a href="https://unite.ut.ee/repository.php">UNITE ITS database</a> can be used as is. To follow along, download the <code>rdp_train_set_16.fa.gz</code> file, and place it in the directory with the fastq files.</p>
<p><strong>Assign taxonomy:</strong></p>
<pre class="r"><code>taxa &lt;- assignTaxonomy(seqtab.nochim, &quot;Training/rdp_train_set_16.fa.gz&quot;, multithread=TRUE)
unname(head(taxa))</code></pre>
<pre><code>##      [,1]       [,2]            [,3]          [,4]           
## [1,] &quot;Bacteria&quot; &quot;Bacteroidetes&quot; &quot;Bacteroidia&quot; &quot;Bacteroidales&quot;
## [2,] &quot;Bacteria&quot; &quot;Bacteroidetes&quot; &quot;Bacteroidia&quot; &quot;Bacteroidales&quot;
## [3,] &quot;Bacteria&quot; &quot;Bacteroidetes&quot; &quot;Bacteroidia&quot; &quot;Bacteroidales&quot;
## [4,] &quot;Bacteria&quot; &quot;Bacteroidetes&quot; &quot;Bacteroidia&quot; &quot;Bacteroidales&quot;
## [5,] &quot;Bacteria&quot; &quot;Bacteroidetes&quot; &quot;Bacteroidia&quot; &quot;Bacteroidales&quot;
## [6,] &quot;Bacteria&quot; &quot;Bacteroidetes&quot; &quot;Bacteroidia&quot; &quot;Bacteroidales&quot;
##      [,5]                 [,6]         
## [1,] &quot;Porphyromonadaceae&quot; &quot;Coprobacter&quot;
## [2,] &quot;Porphyromonadaceae&quot; NA           
## [3,] &quot;Porphyromonadaceae&quot; NA           
## [4,] &quot;Porphyromonadaceae&quot; &quot;Barnesiella&quot;
## [5,] &quot;Bacteroidaceae&quot;     &quot;Bacteroides&quot;
## [6,] &quot;Porphyromonadaceae&quot; NA</code></pre>
<p>Unsurprisingly, the Bacteroidetes are well represented among the most abundant taxa in these fecal samples.</p>
<p><em>The dada2 package implements a <a href="assign.html#species-assignment">second method <code>assignSpecies</code> that uses exact matching to assign 16S sequences to species-level</a> where appropriate.</em></p>
<div style="border: 1px solid red;padding: 5px;background-color: #fff6f6;">
<strong><span style="color:red">If using this workflow on your own data:</span></strong> If your reads do not seem to be appropriately assigned, for example lots of your bacterial 16S sequences are being assigned as <code>Eukaryota NA NA NA NA NA</code>, your reads may be in the opposite orientation as the reference database. Tell dada2 to try the reverse-complement orientation with <code>assignTaxonomy(..., tryRC=TRUE)</code> and see if this fixes the assignments.
</div>
<p> </p>
</div>
<div id="evaluate-accuracy" class="section level1">
<h1>Evaluate accuracy</h1>
<p>One of the processed samples was a “mock community”, in which a mixture of 20 known strains was sequenced (this mock community is supposed to be 21 strains, but <em>P. acnes</em> is absent). Reference sequences corresponding to these strains were provided in the downloaded zip archive. We return to that sample and compare the sequence variants inferred by DADA2 to the expected composition of the community.</p>
<p><strong>Evaluating DADA2’s accuracy on the mock community</strong>:</p>
<pre class="r"><code>unqs.mock &lt;- seqtab.nochim[&quot;Mock&quot;,]
unqs.mock &lt;- sort(unqs.mock[unqs.mock&gt;0], decreasing=TRUE) # Drop ASVs absent in the Mock
cat(&quot;DADA2 inferred&quot;, length(unqs.mock), &quot;sample sequences present in the Mock community.\n&quot;)</code></pre>
<pre><code>## DADA2 inferred 20 sample sequences present in the Mock community.</code></pre>
<pre class="r"><code>mock.ref &lt;- getSequences(file.path(path, &quot;HMP_MOCK.v35.fasta&quot;))
match.ref &lt;- sum(sapply(names(unqs.mock), function(x) any(grepl(x, mock.ref))))
cat(&quot;Of those,&quot;, sum(match.ref), &quot;were exact matches to the expected reference sequences.\n&quot;)</code></pre>
<pre><code>## Of those, 20 were exact matches to the expected reference sequences.</code></pre>
<p>This mock community contained <strong>20</strong> bacterial strains. DADA2 identified <strong>20</strong> sequence variants all of which <strong>exactly</strong> match the reference genomes of the expected community members. The residual error rate after the DADA2 pipeline for this sample is <strong>0%</strong>.</p>
<p>In comparison, <a href="http://www.mothur.org/wiki/MiSeq_SOP#Assessing_error_rates">the mothur pipeline finds 34 OTUs in this Mock community sample</a>. DADA2 infers sequence variants exactly instead of fuzzy 97% OTUs, and outputs fewer false positives to boot!</p>
<p><strong>Here ends the DADA2 portion of the tutorial</strong>.</p>
<hr />
</div>
<div id="bonus-handoff-to-phyloseq" class="section level1">
<h1>Bonus: Handoff to phyloseq</h1>
<p>The DADA2 pipeline produced a sequence table and a taxonomy table which is appropriate for further analysis in phyloseq. We’ll also include the small amount of metadata we have – the samples are named by the gender (G), mouse subject number (X) and the day post-weaning (Y) it was sampled (eg. GXDY).</p>
<p><strong>Import into phyloseq</strong>:</p>
<pre class="r"><code>library(phyloseq); packageVersion(&quot;phyloseq&quot;)</code></pre>
<pre><code>## [1] &#39;1.20.0&#39;</code></pre>
<pre class="r"><code>library(ggplot2); packageVersion(&quot;ggplot2&quot;)</code></pre>
<pre><code>## [1] &#39;2.2.1&#39;</code></pre>
<pre class="r"><code># Make a data.frame holding the sample data
samples.out &lt;- rownames(seqtab.nochim)
subject &lt;- sapply(strsplit(samples.out, &quot;D&quot;), `[`, 1)
gender &lt;- substr(subject,1,1)
subject &lt;- substr(subject,2,999)
day &lt;- as.integer(sapply(strsplit(samples.out, &quot;D&quot;), `[`, 2))
samdf &lt;- data.frame(Subject=subject, Gender=gender, Day=day)
samdf$When &lt;- &quot;Early&quot;
samdf$When[samdf$Day&gt;100] &lt;- &quot;Late&quot;
rownames(samdf) &lt;- samples.out

# Construct phyloseq object (straightforward from dada2 outputs)
ps &lt;- phyloseq(otu_table(seqtab.nochim, taxa_are_rows=FALSE), 
               sample_data(samdf), 
               tax_table(taxa))
ps &lt;- prune_samples(sample_names(ps) != &quot;Mock&quot;, ps) # Remove mock sample
ps</code></pre>
<pre><code>## phyloseq-class experiment-level object
## otu_table()   OTU Table:         [ 229 taxa and 19 samples ]
## sample_data() Sample Data:       [ 19 samples by 4 sample variables ]
## tax_table()   Taxonomy Table:    [ 229 taxa by 6 taxonomic ranks ]</code></pre>
<p>We are now ready to use phyloseq.</p>
<p><strong>Visualize alpha-diversity</strong>:</p>
<pre class="r"><code>plot_richness(ps, x=&quot;Day&quot;, measures=c(&quot;Shannon&quot;, &quot;Simpson&quot;), color=&quot;When&quot;) + theme_bw()</code></pre>
<p><img src="tutorial_files/figure-html/richness-1.png" width="672" /></p>
<p>No obvious systematic difference in alpha-diversity between early and late samples.</p>
<p><strong>Ordinate</strong>:</p>
<pre class="r"><code>ord.nmds.bray &lt;- ordinate(ps, method=&quot;NMDS&quot;, distance=&quot;bray&quot;)</code></pre>
<pre><code>## Square root transformation
## Wisconsin double standardization
## Run 0 stress 0.08908957 
## Run 1 stress 0.1493619 
## Run 2 stress 0.08908959 
## ... Procrustes: rmse 5.693755e-05  max resid 0.0001237885 
## ... Similar to previous best
## Run 3 stress 0.08908957 
## ... New best solution
## ... Procrustes: rmse 9.935625e-06  max resid 2.18974e-05 
## ... Similar to previous best
## Run 4 stress 0.1499185 
## Run 5 stress 0.08908957 
## ... Procrustes: rmse 7.486711e-06  max resid 1.993791e-05 
## ... Similar to previous best
## Run 6 stress 0.3744332 
## Run 7 stress 0.1502671 
## Run 8 stress 0.09001141 
## Run 9 stress 0.09001141 
## Run 10 stress 0.1474728 
## Run 11 stress 0.1556003 
## Run 12 stress 0.161324 
## Run 13 stress 0.08908958 
## ... Procrustes: rmse 4.726698e-05  max resid 9.970712e-05 
## ... Similar to previous best
## Run 14 stress 0.08908957 
## ... Procrustes: rmse 4.270861e-06  max resid 1.052587e-05 
## ... Similar to previous best
## Run 15 stress 0.08908957 
## ... New best solution
## ... Procrustes: rmse 1.263171e-06  max resid 2.65943e-06 
## ... Similar to previous best
## Run 16 stress 0.09001142 
## Run 17 stress 0.08908958 
## ... Procrustes: rmse 3.211466e-05  max resid 6.651034e-05 
## ... Similar to previous best
## Run 18 stress 0.09001141 
## Run 19 stress 0.08908957 
## ... Procrustes: rmse 7.489312e-06  max resid 1.955265e-05 
## ... Similar to previous best
## Run 20 stress 0.09001142 
## *** Solution reached</code></pre>
<pre class="r"><code>plot_ordination(ps, ord.nmds.bray, color=&quot;When&quot;, title=&quot;Bray NMDS&quot;)</code></pre>
<p><img src="tutorial_files/figure-html/ordinate-1.png" width="672" /></p>
<p>Ordination picks out a clear separation between the early and late samples.</p>
<p><strong>Bar plot</strong>:</p>
<pre class="r"><code>top20 &lt;- names(sort(taxa_sums(ps), decreasing=TRUE))[1:20]
ps.top20 &lt;- transform_sample_counts(ps, function(OTU) OTU/sum(OTU))
ps.top20 &lt;- prune_taxa(top20, ps.top20)
plot_bar(ps.top20, x=&quot;Day&quot;, fill=&quot;Family&quot;) + facet_wrap(~When, scales=&quot;free_x&quot;)</code></pre>
<p><img src="tutorial_files/figure-html/bar-plot-1.png" width="672" /></p>
<p>Nothing glaringly obvious jumps out from the taxonomic distribution of the top 20 sequences to explain the early-late differentiation.</p>
<p>This was just a bare bones demonstration of how the data from DADA2 can be easily imported into phyloseq and interrogated. For further examples on the many analyses possible with phyloseq, see <a href="https://joey711.github.io/phyloseq/">the phyloseq web site</a>!</p>
</div>

<div class="container" style="width: 100%;color:grey;text-align:right">
  <hr>
  Maintained by Benjamin Callahan (benjamin DOT j DOT callahan AT gmail DOT com)
</div>



</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
