---
title: "Installation and Testing"
author: "Ben Callahan, Paul McMurdie, Susan Holmes"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteIndexEntry{Installation and Testing}
  \usepackage[utf8]{inputenc}
---


## Installation Instructions

### 0. Install External Dependencies

The following instructions assume that you have installed R 
and have installed 
[Xcode](https://developer.apple.com/xcode/downloads/)
(Mac) or 
[Rtools](http://cran.r-project.org/bin/windows/Rtools/)
(Windows); as well as 
[Bioconductor](http://bioconductor.org/install/).

For Bioconductor,
start a fresh R session, and enter the following.

```{r bioc-install, eval=FALSE}
source("http://bioconductor.org/biocLite.R")
biocLite(suppressUpdates = FALSE)
biocLite("ShortRead", suppressUpdates = FALSE)
```

### 1. Download and unzip.

Download the zipped package (attached) and unzip it.

### 2. Attempt installation from within R

Start a fresh R session, and enter the following.

```{r install-packages, eval=FALSE}
install.packages("path/to/dadac",
                 repos = NULL,
                 type = "source",
                 dependencies = c("Depends", "Suggests","Imports"))
```

For example, if the dadac source code directory was in `~/github/dadac`,
then the following would work.

```{r install-github-example, eval=FALSE}
install.packages("~/github/dadac",
                 repos = NULL,
                 type = "source",
                 dependencies = c("Depends", "Suggests","Imports"))
```

Check Package Version

```{r packageVersion}
packageVersion("dadac")
```


### 3. Troubleshoot Dependencies (Optional)

At this point, 
there may be complaints about missing dependencies.
To install missing dependencies on either
[CRAN](http://cran.r-project.org/)
or
[Bioconductor](http://bioconductor.org/install/), 
start a fresh R session, and enter the following.

```{r bioc-install-missing, eval=FALSE}
source("http://bioconductor.org/biocLite.R")
biocLite("missing_package_1")
biocLite("missing_package_2")
# ... and so on
```

### 4. Re-attempt dadac Installation

Now once again, after dependencies have been installed.

```{r install-packages-rev2, eval=FALSE}
install.packages("path/to/dadac",
                 repos = NULL,
                 type = "source",
                 dependencies = c("Depends", "Suggests","Imports"))
```

You should now be done, if installation was successful.
If not, do panic, and post a message to the issue tracker.



---

## Testing Instructions

### 0. Suggestions

If youâ€™re here, the package installed which is already great! 
Any suggestions or issues with the following 
I would love to hear about.


### 1. Load Package, Explore Function Documentation

```{r load-dadac, message=FALSE}
library("dadac")
```

Now that the package is loaded,
look at the help files for the main implemented functions.

```{r documentation-example, eval=FALSE, message=FALSE}
?dada
?derepFastq
?importUniques
```


### 2. Test out DADA:

THIS IS OLD!

```{r test-dada-0, eval=FALSE}
score <- matrix(c(5, -4, -4, -4, -4, 5, -4, 
                  -4, -4, -4, 5, -4, -4, -4, -4, 5),
                nrow=4, byrow=T)
gap_penalty <- -8
err_init <- matrix(c(0.991, 0.003, 0.003, 0.003, 
                     0.003, 0.991, 0.003, 0.003, 0.003, 
                     0.003, 0.991, 0.003, 0.003, 0.003, 
                     0.003, 0.991),
                   nrow=4, byrow=T)
test_fastq_file = system.file("extdata", "test-nonunique.fastq.gz", package="dadac")
uniques <- derepFastq(fl = test_fastq_file)
uniques.dada <- dada(uniques = uniques$uniques, quals = uniques$quals)
uniques.dada.consistent <- dada(uniques, err_init, score, gap_penalty, self_consist = TRUE)
length(uniques)
length(uniques.dada$genotypes)
length(uniques.dada.consistent$genotypes)
```

```{r}
# Out of `r length(uniques)` original unique sequences,
# did you get `r length(uniques.dada$genotypes)`
# genotypes in `uniques.dada$genotypes` and
# `r length(uniques.dada.consistent$genotypes)` in `uniques.dada.consistent`?
```


### 3. Performance check DADA:

THIS IS OLD!!

```{r compute-performance-1, eval=FALSE}
group_0_25_file = system.file("extdata", "group_0-25.uniques.gz", package="dadac")
bu = importUniques(group_0_25_file)
length(bu)
library("microbenchmark")
microbenchmark(dada(bu, err = err_init, ), times=10)
```

How long does this expression take?
(Hopefully ~10 seconds or less per evaluation.)

THIS IS OLD!

```{r compute-performance-2, eval=FALSE}
group_0_30_file = system.file("extdata", "group_0-30.uniques.gz", package="dadac")
bu = NULL
bu <- importUniques(group_0_30_file)
length(bu)
microbenchmark(dada(bu, err_init, score, gap_penalty), times=10)
```

And this?

### 4. Performance Check

THIS IS OLD!

```{r compute-performance-3, eval=FALSE}
library("ggplot2")
length(bu)
# The max number of sequences from `bu` to include in this example
N = 1000L
# Set random see for repoducibility of example
set.seed(711L)
# Randomly subsample from `bu`, because `bu` 
# on its own is large and takes a long time
buSub0 = bu[sample(x = length(bu), size = N, replace = FALSE)]
# Run and report time elapsed
system.time({
  df <- calibrate_kmers(seqs = names(buSub0),
                        score = get_dada_opt("SCORE_MATRIX"),
                        gap =  -8,
                        max_aligns = 100000)
})
# Make a summary plot
ggplot(df, aes(x=kmer, y=align)) + geom_bin2d()
# Som alternative plots
# ggplot(df, aes(x=kmer, y=align)) + geom_hex()
# ggplot(df, aes(x=kmer, y=align)) + geom_density2d()
```

