# This file was generated by Rcpp::compileAttributes
# Generator token: 10BE3573-1514-4C36-9D1C-5A225CD40393

#' @useDynLib dada2
#' @importFrom Rcpp evalCpp
NULL

dada_uniques <- function(seqs, abundances, err, quals, score, gap, use_kmers, kdist_cutoff, band_size, omegaA, use_singletons, omegaS, max_clust, min_fold, min_hamming, use_quals, qmin, qmax, final_consensus, verbose) {
    .Call('dada2_dada_uniques', PACKAGE = 'dada2', seqs, abundances, err, quals, score, gap, use_kmers, kdist_cutoff, band_size, omegaA, use_singletons, omegaS, max_clust, min_fold, min_hamming, use_quals, qmin, qmax, final_consensus, verbose)
}

C_nwalign <- function(s1, s2, score, gap_p, band) {
    .Call('dada2_C_nwalign', PACKAGE = 'dada2', s1, s2, score, gap_p, band)
}

C_eval_pair <- function(s1, s2) {
    .Call('dada2_C_eval_pair', PACKAGE = 'dada2', s1, s2)
}

C_get_overlaps <- function(s1, s2, allow, max_shift) {
    .Call('dada2_C_get_overlaps', PACKAGE = 'dada2', s1, s2, allow, max_shift)
}

C_pair_consensus <- function(s1, s2) {
    .Call('dada2_C_pair_consensus', PACKAGE = 'dada2', s1, s2)
}

C_pair_consensus2 <- function(s1, s2, prefer) {
    .Call('dada2_C_pair_consensus2', PACKAGE = 'dada2', s1, s2, prefer)
}

#' Generate the kmer-distance and the alignment distance from the
#'   given set of sequences. 
#'
#' @param seqs (Required). Character.
#'  A vector containing all unique sequences in the data set.
#'  Only A/C/G/T allowed.
#'  
#' @param kmer_size (Required). A \code{numeric(1)}. The size of the kmer to test (eg. 5-mer).
#' 
#' @param score (Required). Numeric matrix (4x4).
#' The score matrix used during the alignment. Coerced to integer.
#'
#' @param gap (Required). A \code{numeric(1)} giving the gap penalty for alignment. Coerced to integer.
#'
#' @param band (Required). A \code{numeric(1)} giving the band-size for the NW alignments.
#'
#' @param max_aligns (Required). A \code{numeric(1)} giving the (maximum) number of
#' pairwise alignments to do.
#'
#' @return DataFrame.
#'
#' @export
#' 
#' @examples
#' derep1 = derepFastq(system.file("extdata", "sam1F.fastq.gz", package="dada2"))
#' kmerdf <- evaluate_kmers(getSequences(derep1), 5, getDadaOpt("SCORE_MATRIX"), getDadaOpt("GAP_PENALTY"), 16, 1000)
#' plot(kmerdf$kmer, kmerdf$align)
#' 
evaluate_kmers <- function(seqs, kmer_size, score, gap, band, max_aligns) {
    .Call('dada2_evaluate_kmers', PACKAGE = 'dada2', seqs, kmer_size, score, gap, band, max_aligns)
}

evaluate_band <- function(seqs, score, gap, band_size, max_aligns) {
    .Call('dada2_evaluate_band', PACKAGE = 'dada2', seqs, score, gap, band_size, max_aligns)
}

C_nwvec <- function(s1, s2, score, gap_p, band) {
    .Call('dada2_C_nwvec', PACKAGE = 'dada2', s1, s2, score, gap_p, band)
}

# Register entry points for exported C++ functions
methods::setLoadAction(function(ns) {
    .Call('dada2_RcppExport_registerCCallable', PACKAGE = 'dada2')
})
