# This file was generated by Rcpp::compileAttributes
# Generator token: 10BE3573-1514-4C36-9D1C-5A225CD40393

#' @useDynLib dadac
#' @importFrom Rcpp evalCpp
NULL

#' Run DADA on the provided unique sequences/abundance pairs. 
#'
#' @param seqs (Required). Character.
#'  A vector containing all unique sequences in the data set.
#'  Only A/C/G/T/N/- allowed. Ungapped sequences recommended.
#' 
#' @param abundances (Required). Numeric.
#'  A vector of the number of reads of each unique seuqences.
#'  NAs not tolerated. Must be same length as the seqs vector.
#'
#' @param err (Required). Numeric matrix (4x4).
#'
#' @param score (Required). Numeric matrix (4x4).
#' The score matrix used during the alignment.
#'
#' @param gap (Required). A \code{numeric(1)} giving the gap penalty for alignment.
#'
#' @return DataFrame object with sequence and abundance columns,
#' corresponding to the DADA denoised sample genotypes.
#'
#' @export
dada_uniques <- function(seqs, abundances, err, score, gap) {
    .Call('dadac_dada_uniques', PACKAGE = 'dadac', seqs, abundances, err, score, gap)
}

# Register entry points for exported C++ functions
methods::setLoadAction(function(ns) {
    .Call('dadac_RcppExport_registerCCallable', PACKAGE = 'dadac')
})
