# This file was generated by Rcpp::compileAttributes
# Generator token: 10BE3573-1514-4C36-9D1C-5A225CD40393

#' Exposes ends-free Needleman-Wunsh alignment to R.
#' 
C_nwalign <- function(s1, s2, score, gap_p, band) {
    .Call('dadac_C_nwalign', PACKAGE = 'dadac', s1, s2, score, gap_p, band)
}

#' Calculates the number of matches/mismatches/internal-indels in an alignment.
#' 
C_eval_pair <- function(s1, s2) {
    .Call('dadac_C_eval_pair', PACKAGE = 'dadac', s1, s2)
}

#' Calculates the consensus of two sequences (first sequence wins mismatches).
#' 
C_pair_consensus <- function(s1, s2) {
    .Call('dadac_C_pair_consensus', PACKAGE = 'dadac', s1, s2)
}

#' Generate the kmer-distance and the alignment distance from the
#'   given set of sequences. 
#'
#' @param seqs (Required). Character.
#'  A vector containing all unique sequences in the data set.
#'  Only A/C/G/T allowed.
#' 
#' @param score (Required). Numeric matrix (4x4).
#' The score matrix used during the alignment. Coerced to integer.
#'
#' @param gap (Required). A \code{numeric(1)} giving the gap penalty for alignment. Coerced to integer.
#'
#' @param max_aligns (Required). A \code{numeric(1)} giving the (maximum) number of
#' pairwise alignments to do.
#'
#' @return DataFrame.
#'
#' @export
evaluate_kmers <- function(seqs, kmer_size, score, gap, band, max_aligns) {
    .Call('dadac_evaluate_kmers', PACKAGE = 'dadac', seqs, kmer_size, score, gap, band, max_aligns)
}

#' Quantify the number of alignments altered by banding at the given BAND_SIZE.
#'
#' @param seqs (Required). Character.
#'  A vector containing all unique sequences in the data set.
#'  Only A/C/G/T allowed.
#' 
#' @param score (Required). Numeric matrix (4x4).
#' The score matrix used during the alignment. Coerced to integer.
#'
#' @param gap (Required). A \code{numeric(1)} giving the gap penalty for alignment. Coerced to integer.
#' 
#' @param band_size (Required). A \code{numeric(1)} giving the band size to consider.
#'
#' @param max_aligns (Required). A \code{numeric(1)} giving the (maximum) number of
#' pairwise alignments to do.
#'
#' @return DataFrame.
#'
#' @export
evaluate_band <- function(seqs, score, gap, band_size, max_aligns) {
    .Call('dadac_evaluate_band', PACKAGE = 'dadac', seqs, score, gap, band_size, max_aligns)
}

getSingletonCDF <- function(err, nnt, maxD) {
    .Call('dadac_getSingletonCDF', PACKAGE = 'dadac', err, nnt, maxD)
}

#' @useDynLib dadac
#' @importFrom Rcpp evalCpp
NULL

#' Run DADA on the provided unique sequences/abundance pairs. 
#' 
#' @return List.
#'
#' @export
dada_uniques <- function(seqs, abundances, err, quals, score, gap, use_kmers, kdist_cutoff, band_size, omegaA, use_singletons, omegaS, max_clust, min_fold, min_hamming, use_quals, qmin, qmax, final_consensus, verbose) {
    .Call('dadac_dada_uniques', PACKAGE = 'dadac', seqs, abundances, err, quals, score, gap, use_kmers, kdist_cutoff, band_size, omegaA, use_singletons, omegaS, max_clust, min_fold, min_hamming, use_quals, qmin, qmax, final_consensus, verbose)
}

# Register entry points for exported C++ functions
methods::setLoadAction(function(ns) {
    .Call('dadac_RcppExport_registerCCallable', PACKAGE = 'dadac')
})
